### PR TITLE
Move from CircleCI to Buildkite

### PR DESCRIPTION
**Pull request by mokagio**
What is says in the title. Please forgive the long commits list—This was quite the trial and error. I underestimated the challenge of understanding (part of) `electron-builder` and getting everything it needs to work in a new CI environment.

The latest build shows everything green https://buildkite.com/automattic/simplenote-electron/builds/89

And I used this branch to tag `v2.22.0-beta6` and verify [the GitHub release gets created](https://github.com/Automattic/simplenote-electron/releases/tag/untagged-657b39dc85c254543422).

Unfortunately, the diff is substantial. I left a few comments in there.

### CODE REVIEW

**Comment by mokagio** [comment]
@codebykat @roundhill any ideas why CI would be [failing](https://buildkite.com/automattic/simplenote-electron/builds/16#018fc79a-1994-41ff-b2e6-15cd50190e64) on `make build` with errors like on macOS?

```
[2024-05-30T03:50:44Z] ERROR in ../node_modules/hash-base/node_modules/readable-stream/lib/_stream_readable.js 43:13-37
  | [2024-05-30T03:50:44Z] Module not found: Error: Can't resolve 'buffer' in '/var/lib/buildkite-agent/builds/ci-default-i-0d5406330dcdf6afd-1/automattic/simplenote-electron/node_modules/hash-base/node_modules/readable-stream/lib'
  | [2024-05-30T03:50:44Z]
  | [2024-05-30T03:50:44Z] BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
  | [2024-05-30T03:50:44Z] This is no longer the case. Verify if you need this module and configure a polyfill for it.
  | [2024-05-30T03:50:44Z]
  | [2024-05-30T03:50:44Z] If you want to include a polyfill, you need to:
  | [2024-05-30T03:50:44Z] 	- add a fallback 'resolve.fallback: { "buffer": require.resolve("buffer/") }'
  | [2024-05-30T03:50:44Z] 	- install 'buffer'
  | [2024-05-30T03:50:44Z] If you don't want to include a polyfill, you can use an empty module like this:
  | [2024-05-30T03:50:44Z] 	resolve.fallback: { "buffer": false }
  | [2024-05-30T03:50:44Z]  @ ../node_modules/hash-base/node_modules/readable-stream/readable-browser.js 1:10-63
  | [2024-05-30T03:50:44Z]  @ ../node_modules/hash-base/index.js 3:16-52
  | [2024-05-30T03:50:44Z]  @ ../node_modules/md5.js/index.js 3:15-35
  | [2024-05-30T03:50:44Z]  @ ./dialogs/share/index.tsx 4:0-25 40:21-24
  | [2024-05-30T03:50:44Z]  @ ./dialog-renderer/index.tsx 12:0-43 47:68-79
  | [2024-05-30T03:50:44Z]  @ ./app.tsx 10:0-47 159:303-317
  | [2024-05-30T03:50:44Z]  @ ./boot-with-auth.tsx 5:0-24 40:40-43
  | [2024-05-30T03:50:44Z]  @ ./boot.ts
```

On my machine, it works fine 🤔 Thanks!

---

**Update** a6613931fea70001bd2a7db60f590fb2635c90c7 fixed the issue 👍

File: .buildkite/commands/package_windows.ps1
Comment (mokagio) [inline .buildkite/commands/package_windows.ps1:19]: I need to followup on whether `$env:<VAR NAME>` would be better here and save the trouble of doing multiple checks.

File: .buildkite/commands/package_windows.ps1
Comment (mokagio) [inline .buildkite/commands/package_windows.ps1:4]: This comes from the CI Toolkit Buildkite plugin.

File: .buildkite/commands/package_windows.ps1
Comment (mokagio) [inline .buildkite/commands/package_windows.ps1:42]: I suppose we could shell out and call `install_node_dependencies.sh` here too 🤔

File: .buildkite/pipeline.yml
Comment (mokagio) [inline .buildkite/pipeline.yml:63]: I considered running `make build` standalone and then download the artifacts, but the overhead at buildtime is minimal compared to the rest of the step.

File: .buildkite/pipeline.yml
Comment (mokagio) [inline .buildkite/pipeline.yml]:
```suggestion
      # The behavior occurs in CircleCI, too. See:
```

File: .buildkite/shared-pipeline-vars
Comment (mokagio) [inline .buildkite/shared-pipeline-vars:8]: This will eventually land in the CI Toolkit `main`. I'll followup once that happens. There are some issues on that end with how the tests are written.

File: .configure
Comment (mokagio) [inline .configure:8]: Unfortunately, the existing notarization script requires the `p8` file and does not support reading the key content from memory like Fastlane does. As such, we need to track the encrypted `p8` in the repo.

File: Makefile
Comment (mokagio) [inline Makefile:123]: These verbose logs helped me get my head around what was going on at call time.

File: Makefile
Comment (mokagio) [inline Makefile:146]: CircleCI had two steps to build either `exe` or `appx`. I thought it made more sense to build everything in one go, in particular given the boot overhead for Windows CI machines.

File: after_sign_hook.js
Comment (mokagio) [inline after_sign_hook.js:52]: Interestingly, we are not getting a lint issue for this one, like the one above 🤷‍♂️

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile]:
```suggestion
```

File: fastlane/Fastfile
Comment (mokagio) [inline fastlane/Fastfile:65]: Eventually, it would be good to use Fastlane for notarization, too. It's a better toolchain and used by more Apple developers than the Electron subset.

**Review by codebykat — APPROVED** [review]
Ooooof. This has been truly a nightmare and I know I haven't been much help due to my unfamiliarity with Buildkite and the current mobile deployment processes. Thanks so much for spending the time to figure it out!

Most of this code makes my brain hurt, but I've read through it and it seems to make sense. I trust your judgment (and the passing builds!). Kind of ridiculous that we are now adding a Gemfile on top of the node package lockfile... lol, software ecosystems are all terrible. 🫠 

I did test the built artifact on OSX and it is working well. I wasn't able to get a Linux or Windows VM working, not for lack of trying. Hopefuly @roundhill can.

File: .buildkite/pipeline.yml
Comment (codebykat) [inline .buildkite/pipeline.yml]: This should write to `config.json`, not `config-local.json`. Doing that should fix the ability to log in, I hope 🤞

Comment (mokagio) [reply .buildkite/pipeline.yml]: Ah, good point!

`Makefile` makes it go to `config-local.json` but that must because it's intended for local use

https://github.com/Automattic/simplenote-electron/blob/f5d463f7a5f5beb3638696c9700b1c7781448744/Makefile#L187-L193

**Comment by mokagio** [comment]
> Ooooof. This has been truly a nightmare and I know I haven't been much help due to my unfamiliarity with Buildkite and the current mobile deployment processes. [...]
>
> Most of this code makes my brain hurt, but I've read through it and it seems to make sense. I trust your judgment (and the passing builds!). Kind of ridiculous that we are now adding a Gemfile on top of the node package lockfile... lol, software ecosystems are all terrible. 🫠

Thanks a lot @codebykat !

I should have asked a review from my team at @Automattic/apps-infrastructure at the same time as you. That's more our domain after all 😳 🙇‍♂️  Updated now.

File: .buildkite/pipeline.yml
Comment (mokagio) [reply .buildkite/pipeline.yml]: See https://github.com/Automattic/simplenote-electron/pull/3218 as an attempt to address this in a DRY way.

**Comment by AliSoftware** [comment]
Sorry for the delay in getting to this @mokagio 😓  I kept postponing the review because the diff was a bit big and wanted to find a moment when I'd be able to take the time… and thus planned to review it on my today or tomorrow (i.e. once I'd have finished my current high priority tasks of the past few days) 😓 .  But I guess as long as the CI passes on Buildkite, there's not much to review anyway 🙃

File: .buildkite/commands/package_windows.ps1
Comment (AliSoftware) [reply .buildkite/commands/package_windows.ps1:4]: Thanks for documenting that. Wonder if that would be worth a code comment?

File: .buildkite/commands/package_windows.ps1
Comment (AliSoftware) [inline .buildkite/commands/package_windows.ps1:8]: Gosh this Powershell syntax… 😅  I'm surprised there might not be a simpler way to read an env var (whatever the level in which it's defined, process or machine) 🤔 

[EDIT] Oh, [I see you already have this on your radar](https://github.com/Automattic/simplenote-electron/pull/3204#discussion_r1637742425) 👍

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:2]: Is this still needed, or was this only added during debugging but could be removed now?

File: .buildkite/pipeline.yml
Comment (AliSoftware) [inline .buildkite/pipeline.yml:35]:
```suggestion
  # It's up to the elector-builder configuration to decide whether to upload
```

File: .buildkite/pipeline.yml
Comment (AliSoftware) [reply .buildkite/pipeline.yml:2]: Oh I see I guess this is related to https://github.com/Automattic/simplenote-electron/pull/3204/files#r1637752108 and will be removed once you land that to `a8c-ci-toolkit`'s trunk` 👍

**Pull request merged**
